### ISSUE TITLE
Exception : Collection has changed while changing it 

### ISSUE THREAD

**Issue by m4kz** [mxjones/RedBus#1]
Hi,

Thanks for your work, but there is an issue in EventBus, where you use a for-each to alter a collection.
I have fixed it by using a normal for instead:

```
    `public void Publish<TEventBase>(TEventBase eventItem) where TEventBase : EventBase
    {
        if (eventItem == null)
            throw new ArgumentNullException("eventItem");

        List<ISubscription> allSubscriptions = new List<ISubscription>();
        lock (SubscriptionsLock)
        {
            if (_subscriptions.ContainsKey(typeof(TEventBase)))
                allSubscriptions = _subscriptions[typeof(TEventBase)];
        }

        for (int listIndex = allSubscriptions.Count - 1; listIndex >= 0; listIndex--)
        {
            try
            {
                allSubscriptions[listIndex].Publish(eventItem);
            }
            catch (Exception exception)
            {
            }
        }
    }`
```

Regards :)

**Comment by mxjones** [comment]
Hi,

Thanks for raising this - I haven't seen this issue come up before, do you have a reliable way/example code to reproduce this so I could see it in action?

Thanks for posting your fix, I'll look at getting that put in & a new Nuget package published!

Cheers

**Comment by m4kz** [comment]
I have wrapped Eventbus in my own class so I don't have to propagate RedBus code in mine.
I think the problem might have been triggered by the lock around. 
As a separated question, Whats the license of Redbus?

`public void Publish<TEvent, TView>(TEvent myEvent)
            where TView : IMyView
            where TEvent : IMyEvent<TView>
        {
            lock (eventRegistrationLock)
            {
                PayloadEvent<TEvent> payLoadEvent = new PayloadEvent<TEvent>(myEvent);
                eventBus.Publish(payLoadEvent);
            }
        }
`

**Comment by mxjones** [comment]
Hi,

Ah ok - I am currently doing the same as you except I am not using a lock, if I get some time I will try and reproduce this issue this week.

Hadn't thought of License - MIT / Feel free to use it however and wherever you want, I will add a license to the repo later on.

Cheers

**Comment by mxjones** [comment]
Apologies for taking so long on this one - just pushed the fix to master now. New NuGet package is available aswell (1.0.4) - Cheers!
